packages/go_router/lib/src/parser.dart

```diff
@@ -240,3 +343,303 @@ class GoRouteInformationParser extends RouteInformationParser<RouteMatchList> {
+    // `routeInformation` has already been normalized by the parser entrypoint.
+    final RouteMatchList? redirectionErrorMatchList =
+        _redirectionErrorMatchList(context, routeInformation.uri, infoState);
+
+    if (redirectionErrorMatchList != null) {
+      // Return immediately if the redirection limit is exceeded.
+      return SynchronousFuture<RouteMatchList>(redirectionErrorMatchList);
+    }
+
+    // Find route matches for the normalized URI.
+    final RouteMatchList incomingMatches = _configuration.findMatch(
+      routeInformation.uri,
+      extra: infoState.extra,
+    );
+
+    // Build the next navigation state.
+    final GoRouterState nextState = _buildTopLevelGoRouterState(
+      incomingMatches,
+    );
+
+    // Get the current state from the router delegate.
+    final RouteMatchList currentMatchList =
+        _router.routerDelegate.currentConfiguration;
+    final GoRouterState currentState =
+        currentMatchList.isNotEmpty
+            ? _buildTopLevelGoRouterState(currentMatchList)
+            : nextState;
+
+    // Execute the onEnter callback in a try-catch to capture synchronous exceptions.
+    Future<OnEnterResult> onEnterResultFuture;
+    try {
+      final FutureOr<OnEnterResult> result = topOnEnter(
+        context,
+        currentState,
+        nextState,
+        _router,
+      );
+      // Convert FutureOr to Future
+      onEnterResultFuture =
+          result is OnEnterResult
+              ? SynchronousFuture<OnEnterResult>(result)
+              : result;
+    } catch (error) {
+      final RouteMatchList errorMatchList = _errorRouteMatchList(
+        routeInformation.uri,
+        error is GoException ? error : GoException(error.toString()),
+        extra: infoState.extra,
+      );
+
+      _resetRedirectionHistory();
+
+      return SynchronousFuture<RouteMatchList>(
+        _onParserException != null
+            ? _onParserException(context, errorMatchList)
+            : errorMatchList,
+      );
+    }
+
+    // Handle asynchronous completion and catch any errors.
+    return onEnterResultFuture.then<RouteMatchList>(
+      (OnEnterResult result) async {
+        RouteMatchList matchList;
+        final OnEnterThenCallback? callback = result.then;
+
+        if (result is Allow) {
+          matchList = await onCanEnter();
+          _resetRedirectionHistory(); // reset after committed navigation
+        } else {
+          // Block: check if this is a hard stop or chaining block
+          log(
+            'onEnter blocked navigation from ${currentState.uri} to ${nextState.uri}',
+          );
+          matchList = await onCanNotEnter();
+
+          // Treat `Block.stop()` as the explicit hard stop.
+          // We intentionally don't try to detect "no-op" callbacks; any
+          // Block with `then` keeps history so chained guards can detect loops.
+          final bool hardStop = result is Block && callback == null;
```

Comment:
consider add a bool getter isStop on Block class so that we don't infer from callback== null